## Data Understanding
For this analysis, we will utilize a database from BLAH BLAH BLAH. It contains BLAH BLAH BLAH, Describe dataset.


In [1]:
#import necessary packages
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder, LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.datasets import make_regression
from sklearn.linear_model import LinearRegression
import sklearn.metrics as metrics
from random import gauss
from mpl_toolkits.mplot3d import Axes3D
from scipy import stats as stats
from sklearn.feature_selection import RFE


%matplotlib inline

In [2]:
#ignore pairplot and graph warnings
import warnings
warnings.filterwarnings('ignore')

In [3]:
housing = pd.read_csv('data/kc_house_data.csv')
housing.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0      5650     1.0        NaN  NONE  ...      7 Average       1180   
1      7242     2.0         NO  NONE  ...      7 Average       2170   
2     10000     1.0         NO  NONE  ...  6 Low Average        770   
3      5000     1.0         NO  NONE  ...      7 Average       1050   
4      8080     1.0         NO  NONE  ...         8 Good       1680   

   sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0            0.0     1955           0.0    98178  47.5112 -122.257   
1          400.0     1951        1991.0    98125  47.7210 -122.319   
2            0.0     1933           NaN    98028  47.7379 -122.233   
3          910.0     1965           0.0    98136  47.5208 -122.393   
4            0.0     1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 21 columns]

In [4]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

In [5]:
housing.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    sqft_above      yr_built  yr_renovated  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  17755.000000   
mean   1.509941e+04      1.494096   1788.596842   1970.999676     83.636778   
std    4.141264e+04      0.539683    827.759761     29.375234    399.946414   
min    5.200000e+02      1.000000    370.000000   1900.000000      0.000000   
25%    5.040000e+03      1.000000   1190.000000   1951.000000      0.000000   
50%    7.618000e+03      1.500000   1560.000000   1975.000000      0.000000   
75%    1.068500e+04      2.000000   2210.000000   1997.000000      0.000000   
max    1.651359e+06      3.500000   9410.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000  21597.000000   21597.000000   21597.000000  
mean   98077.951845     47.560093   -122.213982    1986.620318   12758.283512  
std       53.513072      0.138552      0.140724     685.230472   27274.441950  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471100   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.231000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [6]:
housing.isnull().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

Observations: We have several columns with categorical values
- waterfront      
- view             
- condition       
- grade          
- floors 
- bedrooms
- bathrooms
- zipcode

Also the column sqft_basement is an object. Let's find out why.

We have several null values in the waterfont and yr_renovated categories. we have a small number of null values in the view category. Yr_renovated seems self-explanatory - a nan value means that the
house was not renovated. Let's explore the remaining two categories. 


**View**

- several options - none, nan, good, excellent, average, and fair.

We can either assume a nan response means no view and replace all the nan values with none, 
or we can just drop the rows with a nan response.  
Since our data comprises of ~21,600 different rows, if we dropped these 63 rows, 
we'd only be loosing 0.3% of our data.

Let's keep them and assume a nan response means no view.

**Sqft_basement**
- let's drop the one row where the sqft_basement is a ?. 

## Data Preparation / Cleaning

- figuring out how to replace null values in columns
- investigating duplicate id
- creating null columns

In [13]:
housing['date'] =  pd.to_datetime(housing['date'])
housing

id       date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520 2014-10-13  221900.0         3       1.00         1180   
1      6414100192 2014-12-09  538000.0         3       2.25         2570   
2      5631500400 2015-02-25  180000.0         2       1.00          770   
3      2487200875 2014-12-09  604000.0         4       3.00         1960   
4      1954400510 2015-02-18  510000.0         3       2.00         1680   
...           ...        ...       ...       ...        ...          ...   
21592   263000018 2014-05-21  360000.0         3       2.50         1530   
21593  6600060120 2015-02-23  400000.0         4       2.50         2310   
21594  1523300141 2014-06-23  402101.0         2       0.75         1020   
21595   291310100 2015-01-16  400000.0         3       2.50         1600   
21596  1523300157 2014-10-15  325000.0         2       0.75         1020   

       sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0          5650     1.0        NaN  NONE  ...      7 Average       1180   
1          7242     2.0         NO  NONE  ...      7 Average       2170   
2         10000     1.0         NO  NONE  ...  6 Low Average        770   
3          5000     1.0         NO  NONE  ...      7 Average       1050   
4          8080     1.0         NO  NONE  ...         8 Good       1680   
...         ...     ...        ...   ...  ...            ...        ...   
21592      1131     3.0         NO  NONE  ...         8 Good       1530   
21593      5813     2.0         NO  NONE  ...         8 Good       2310   
21594      1350     2.0         NO  NONE  ...      7 Average       1020   
21595      2388     2.0        NaN  NONE  ...         8 Good       1600   
21596      1076     2.0         NO  NONE  ...      7 Average       1020   

       sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0                0.0     1955           0.0    98178  47.5112 -122.257   
1              400.0     1951        1991.0    98125  47.7210 -122.319   
2                0.0     1933           NaN    98028  47.7379 -122.233   
3              910.0     1965           0.0    98136  47.5208 -122.393   
4                0.0     1987           0.0    98074  47.6168 -122.045   
...              ...      ...           ...      ...      ...      ...   
21592            0.0     2009           0.0    98103  47.6993 -122.346   
21593            0.0     2014           0.0    98146  47.5107 -122.362   
21594            0.0     2009           0.0    98144  47.5944 -122.299   
21595            0.0     2004           0.0    98027  47.5345 -122.069   
21596            0.0     2008           0.0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357  

[21597 rows x 21 columns]

In [14]:
housing['id'].value_counts()

795000620     3
1825069031    2
2019200220    2
7129304540    2
1781500435    2
             ..
7812801125    1
4364700875    1
3021059276    1
880000205     1
1777500160    1
Name: id, Length: 21420, dtype: int64

In [15]:
idcheck1 = housing[housing['id'] == 795000620]
idcheck1

id       date     price  bedrooms  bathrooms  sqft_living  \
17588  795000620 2014-09-24  115000.0         3        1.0         1080   
17589  795000620 2014-12-15  124000.0         3        1.0         1080   
17590  795000620 2015-03-11  157000.0         3        1.0         1080   

       sqft_lot  floors waterfront  view  ...   grade sqft_above  \
17588      6250     1.0         NO  NONE  ...  5 Fair       1080   
17589      6250     1.0         NO  NONE  ...  5 Fair       1080   
17590      6250     1.0        NaN  NONE  ...  5 Fair       1080   

       sqft_basement yr_built  yr_renovated  zipcode      lat    long  \
17588            0.0     1950           0.0    98168  47.5045 -122.33   
17589            0.0     1950           0.0    98168  47.5045 -122.33   
17590            0.0     1950           NaN    98168  47.5045 -122.33   

       sqft_living15  sqft_lot15  
17588           1070        6250  
17589           1070        6250  
17590           1070        6250  

[3 rows x 21 columns]

In [16]:
idcheck2 = housing[housing['id'] == 1825069031]
idcheck2

id       date     price  bedrooms  bathrooms  sqft_living  \
3946  1825069031 2014-08-14  550000.0         4       1.75         2410   
3947  1825069031 2014-10-16  550000.0         4       1.75         2410   

      sqft_lot  floors waterfront  view  ...   grade sqft_above  \
3946      8447     2.0         NO  GOOD  ...  8 Good       2060   
3947      8447     2.0        NaN  GOOD  ...  8 Good       2060   

      sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
3946          350.0     1936        1980.0    98074  47.6499 -122.088   
3947          350.0     1936        1980.0    98074  47.6499 -122.088   

      sqft_living15  sqft_lot15  
3946           2520       14789  
3947           2520       14789  

[2 rows x 21 columns]

In [17]:
housing['waterfront'].unique()

array([nan, 'NO', 'YES'], dtype=object)

In [23]:
view_waterfront = housing[(housing['waterfront'].isna()) & (housing['view'] == "NONE")]
view_waterfront

id       date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520 2014-10-13  221900.0         3       1.00         1180   
10     1736800520 2015-04-03  662500.0         3       2.50         3560   
23     8091400200 2014-05-16  252700.0         2       1.50         1070   
40     5547700270 2014-07-15  625000.0         4       2.50         2570   
55     9822700295 2014-05-12  885000.0         4       2.50         2830   
...           ...        ...       ...       ...        ...          ...   
21567  2025049203 2014-06-10  399950.0         2       1.00          710   
21578  5087900040 2014-10-17  350000.0         4       2.75         2500   
21586   844000965 2014-06-26  224000.0         3       1.75         1500   
21587  7852140040 2014-08-25  507250.0         3       2.50         2270   
21595   291310100 2015-01-16  400000.0         3       2.50         1600   

       sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0          5650     1.0        NaN  NONE  ...      7 Average       1180   
10         9796     1.0        NaN  NONE  ...         8 Good       1860   
23         9643     1.0        NaN  NONE  ...      7 Average       1070   
40         5520     2.0        NaN  NONE  ...       9 Better       2570   
55         5000     2.0        NaN  NONE  ...       9 Better       2830   
...         ...     ...        ...   ...  ...            ...        ...   
21567      1157     2.0        NaN  NONE  ...      7 Average        710   
21578      5995     2.0        NaN  NONE  ...         8 Good       2500   
21586     11968     1.0        NaN  NONE  ...  6 Low Average       1500   
21587      5536     2.0        NaN  NONE  ...         8 Good       2270   
21595      2388     2.0        NaN  NONE  ...         8 Good       1600   

       sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0                0.0     1955           0.0    98178  47.5112 -122.257   
10            1700.0     1965           0.0    98007  47.6007 -122.145   
23               0.0     1985           NaN    98030  47.3533 -122.166   
40               0.0     2000           NaN    98074  47.6145 -122.027   
55               0.0     1995           0.0    98105  47.6597 -122.290   
...              ...      ...           ...      ...      ...      ...   
21567            0.0     1943           0.0    98102  47.6413 -122.329   
21578            0.0     2008           0.0    98042  47.3749 -122.107   
21586            0.0     2014           0.0    98010  47.3095 -122.002   
21587            0.0     2003           0.0    98065  47.5389 -121.881   
21595            0.0     2004           0.0    98027  47.5345 -122.069   

       sqft_living15  sqft_lot15  
0               1340        5650  
10              2210        8925  
23              1220        8386  
40              2470        5669  
55              1950        5000  
...              ...         ...  
21567           1370        1173  
21578           2530        5988  
21586           1320       11303  
21587           2270        5731  
21595           1410        1287  

[2110 rows x 21 columns]

In [ ]:
housing['waterfront'].fillna(value='NO', inplace=True)

**Waterfront view**

- the only options are yes or no. 
Any property with a waterfront will have a non-NONE view. Since there are 2110 NONE views in our NA waterfront set, it seems safer to assume NO as the default for waterfront.


We can either assume no waterfront view and replace all the nan values with no, or we can just drop the rows with a 
nan response. 
Since our data comprises of ~21,600 different rows, if we dropped these 2376 rows, we'd be loosing 
more than 10% of our data. 

So let's keep them and assume a nan response means no waterfront view. 

In [18]:
housing['sqft_basement'].value_counts()

0.0       12826
?           454
600.0       217
500.0       209
700.0       208
          ...  
2400.0        1
2570.0        1
875.0         1
2810.0        1
2580.0        1
Name: sqft_basement, Length: 304, dtype: int64

In [ ]:
#housing.drop(['id','date', 'lat', 'long'], axis=1, inplace=True)

In [19]:
housing['view'].unique()

array(['NONE', nan, 'GOOD', 'EXCELLENT', 'AVERAGE', 'FAIR'], dtype=object)

In [20]:
housing['yr_renovated'].unique()

array([   0., 1991.,   nan, 2002., 2010., 1992., 2013., 1994., 1978.,
       2005., 2003., 1984., 1954., 2014., 2011., 1983., 1945., 1990.,
       1988., 1977., 1981., 1995., 2000., 1999., 1998., 1970., 1989.,
       2004., 1986., 2007., 1987., 2006., 1985., 2001., 1980., 1971.,
       1979., 1997., 1950., 1969., 1948., 2009., 2015., 1974., 2008.,
       1968., 2012., 1963., 1951., 1962., 1953., 1993., 1996., 1955.,
       1982., 1956., 1940., 1976., 1946., 1975., 1964., 1973., 1957.,
       1959., 1960., 1967., 1965., 1934., 1972., 1944., 1958.])

In [ ]:
na_waterfront = housing[housing['waterfront'].isna()]
na_waterfront


In [ ]:
housing['view'].fillna(value='NONE', inplace=True)

In [ ]:
housing.head(10)

In [ ]:
housing.columns

In [ ]:
#range 1-11. figure out 33 either use average after train/test split. or assume 33 is typo for 3
bedrooms = housing['bedrooms'].unique()
bathrooms = housing['bathrooms'].unique()
floors = housing['floors'].unique()
print (f"This is the list of bedrooms {bedrooms}")
print (f"This is the list of bathrooms {bathrooms}")
print (f"This is the list of floors {floors}")

In [ ]:
housing['lat'].min()

In [ ]:
housing['zipcode'].value_counts()

In [ ]:
#Setting up train test split
X = housing.drop('price', axis=1)
y = housing['price']

X_train , X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=57)

In [ ]:
Housing_train= 

In [ ]:
lb_make = Labelencoder()


EXPLORATORY

In [ ]:
sns.pairplot(housing)
plt.show()

In [24]:
housing.corr()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id             1.000000 -0.016772  0.001150   0.005162    -0.012241 -0.131911   
price         -0.016772  1.000000  0.308787   0.525906     0.701917  0.089876   
bedrooms       0.001150  0.308787  1.000000   0.514508     0.578212  0.032471   
bathrooms      0.005162  0.525906  0.514508   1.000000     0.755758  0.088373   
sqft_living   -0.012241  0.701917  0.578212   0.755758     1.000000  0.173453   
sqft_lot      -0.131911  0.089876  0.032471   0.088373     0.173453  1.000000   
floors         0.018608  0.256804  0.177944   0.502582     0.353953 -0.004814   
sqft_above    -0.010799  0.605368  0.479386   0.686668     0.876448  0.184139   
yr_built       0.021617  0.053953  0.155670   0.507173     0.318152  0.052946   
yr_renovated  -0.012010  0.129599  0.018495   0.051050     0.055660  0.004513   
zipcode       -0.008211 -0.053402 -0.154092  -0.204786    -0.199802 -0.129586   
lat           -0.001798  0.306692 -0.009951   0.024280     0.052155 -0.085514   
long           0.020672  0.022036  0.132054   0.224903     0.241214  0.230227   
sqft_living15 -0.002701  0.585241  0.393406   0.569884     0.756402  0.144763   
sqft_lot15    -0.138557  0.082845  0.030690   0.088303     0.184342  0.718204   

                 floors  sqft_above  yr_built  yr_renovated   zipcode  \
id             0.018608   -0.010799  0.021617     -0.012010 -0.008211   
price          0.256804    0.605368  0.053953      0.129599 -0.053402   
bedrooms       0.177944    0.479386  0.155670      0.018495 -0.154092   
bathrooms      0.502582    0.686668  0.507173      0.051050 -0.204786   
sqft_living    0.353953    0.876448  0.318152      0.055660 -0.199802   
sqft_lot      -0.004814    0.184139  0.052946      0.004513 -0.129586   
floors         1.000000    0.523989  0.489193      0.003535 -0.059541   
sqft_above     0.523989    1.000000  0.424037      0.022137 -0.261570   
yr_built       0.489193    0.424037  1.000000     -0.225209 -0.347210   
yr_renovated   0.003535    0.022137 -0.225209      1.000000  0.069846   
zipcode       -0.059541   -0.261570 -0.347210      0.069846  1.000000   
lat            0.049239   -0.001199 -0.148370      0.032145  0.266742   
long           0.125943    0.344842  0.409993     -0.072156 -0.564259   
sqft_living15  0.280102    0.731767  0.326377     -0.000247 -0.279299   
sqft_lot15    -0.010722    0.195077  0.070777      0.003852 -0.147294   

                    lat      long  sqft_living15  sqft_lot15  
id            -0.001798  0.020672      -0.002701   -0.138557  
price          0.306692  0.022036       0.585241    0.082845  
bedrooms      -0.009951  0.132054       0.393406    0.030690  
bathrooms      0.024280  0.224903       0.569884    0.088303  
sqft_living    0.052155  0.241214       0.756402    0.184342  
sqft_lot      -0.085514  0.230227       0.144763    0.718204  
floors         0.049239  0.125943       0.280102   -0.010722  
sqft_above    -0.001199  0.344842       0.731767    0.195077  
yr_built      -0.148370  0.409993       0.326377    0.070777  
yr_renovated   0.032145 -0.072156      -0.000247    0.003852  
zipcode        0.266742 -0.564259      -0.279299   -0.147294  
lat            1.000000 -0.135371       0.048679   -0.086139  
long          -0.135371  1.000000       0.335626    0.255586  
sqft_living15  0.048679  0.335626       1.000000    0.183515  
sqft_lot15    -0.086139  0.255586       0.183515    1.000000

In [ ]:
housing.corr().price.sort_values(ascending=False)

In [ ]:
Simple Model - no encoding

In [ ]:
housing['sqft_above'].sort_values()

In [ ]:
#Multiple Regression model all variables (no encoding)
housing1 = housing.drop(['id','date', 'grade', 'condition', 'waterfront', 'view', 'sqft_basement', 'bedrooms', 'bathrooms', 'floors', 'yr_renovated', 'zipcode', 'lat', 'long'], axis=1)
housing1.head()

In [ ]:
lr_rfe = LinearRegression()
select = RFE(lr_rfe, n_features_to_select =4)

In [ ]:
ss = StandardScaler()
ss.fit(housing1.drop('price', axis=1))
housing1_scaled = ss.transform(housing1.drop('price', axis=1))

In [ ]:
housing1_scaled

In [ ]:
select.fit(X=housing1_scaled, y=housing1['price'])

In [ ]:
select.support_

In [ ]:
select.ranking_

In [ ]:
## use sqft_living  yr_built  sqft_living15  sqft_lot15

In [ ]:
Polynomials using all except categorical values

In [ ]:
X = housing1.drop('price', axis=1)
y = housing1['price']
pf = PolynomialFeatures(degree=3)
pf.fit(X)

In [ ]:
pf.transform(X)

In [ ]:
pf.transform(X).shape

In [ ]:
pf.get_feature_names()

In [ ]:
polynomial_df = pd.DataFrame(pf.transform(X), columns= pf.get_feature_names() )

In [ ]:
lr = LinearRegression()
lr.fit(polynomial_df, y)

In [ ]:
lr.score(polynomial_df, y)

## DO NOT RUN CELL BELOW 

In [ ]:
sns.pairplot(housing, 'price')

In [ ]:
housing.corr().price.sort_values(ascending=False)

In [ ]:
mask = np.triu(np.ones_like(housing.corr(), dtype=bool))

plt.figure(figsize=(20,10))

cor = housing.corr().abs()
sns.heatmap(cor, mask=mask, annot=True);